pandas/tests/test_index.py

```diff
@@ -1909,6 +1909,46 @@ def test_astype(self):
         self.assertTrue(i.equals(result))
         self.check_is_index(result)
 
+    def test_promote_type(self):
+        # GH-9966
+        index_int = Int64Index(np.arange(5))
+        operators = (operator.add,
+                     operator.sub,
+                     operator.mul,
+                     operator.truediv,
+                     operator.floordiv)
+        for op in operators:
+            index_promoted = op(index_int, np.pi)
+            self.assertEqual(index_promoted.dtype, np.float64)
+            self.assertIsInstance(index_promoted, Float64Index)
+
+    def test_promote_type_inplace(self):
+        # Related to GH-9966
+        index = Int64Index(np.arange(5))
+        index += np.pi
```

Comment:
I know that in some cases Python will automatically treat an in-place operation like `x += y` as `x = x + y`, e.g., if you implement `__add__` but not `__iadd__`. This is how inplace operations work on integers, for example, which are immutable.

In any case, this should certainly fail with `TypeError` if `index` is the same object on both these lines.